ipn/ipnlocal/serve.go

```diff
@@ -398,14 +398,19 @@ func (b *LocalBackend) tcpHandlerForServe(dport uint16, srcAddr netip.AddrPort)
 	return nil
 }
 
+func getServeHTTPContext(r *http.Request) (c *serveHTTPContext, ok bool) {
+	c, ok = r.Context().Value(serveHTTPContextKey{}).(*serveHTTPContext)
+	return c, ok
```

Comment:
two-three line functions is one of the few times naked returns are permissible if you prefer. either way.